### DISCUSSION TITLE
Packages referencing each other

### DISCUSSION THREAD

**Discussion by mweel1** [vercel/turborepo#4224] [Help] [answered]
I am a bit stuck.  In my packages I have two workspaces one is a message context in react that display messages to users with some toast function.  My other workspace is a useFetch that calles API's.

If the the fetch gets a 500 error I want to display a message using the message context workspace.

However, my components both use each other components can fetch data, and now my fetch can use components.

When I try to reference each other I get an error around circular dependencies.

Wanted to discuss some ideas around this problem?

The only thing I can think of is to put everything in one workspace in package.   Could I do multiple exports and have "ui/components", "ui/hooks"?

Use events between components (seems a bit hackish).

This seems like a pretty normal use case to share libraries in your packages workspaces.

Thoughts?

Thanks!!

**Comment by anthonyshew** [comment, marked as answer]
Turborepo uses a directed **acyclic** graph under the hood, meaning you won't be able to create cycling dependencies, as you've found. For more what we do to build your Task Graph, check out this doc: https://turbo.build/repo/docs/core-concepts/monorepos/task-graph

You do have options to get this problem solved:
- As you've described, you can have multiple exports from one package. The `ui` package in `npx create-turbo@latest` has an example of this in `packages/ui/package.json`.
- If your requirements mean you really do want/need two different workspaces, you can create a third workspace that handles the intersection of the two dependent packages and use that third package where needed.
- If you still find you can't get rid of cyclic dependencies, you may want to have a rethink at your packaging architecture. Cyclical dependencies are generally problematic (both in and out of the context of `turbo`) because you can't count on the behavior of the cycle. Often, you'll find you really *can* break the cycle upon further inspection.
